blockchain/chain/src/store/chain_store.rs

```diff
@@ -300,23 +280,85 @@ where
     Ok(())
 }
 
-/// Gets 32 bytes of randomness for ChainRand paramaterized by the DomainSeparationTag, ChainEpoch, Entropy
-pub fn get_randomness<DB: BlockStore>(
+/// Finds the latest beacon entry given a tipset up to 20 blocks behind
+pub fn latest_beacon_entry<DB>(db: &DB, ts: &Tipset) -> Result<BeaconEntry, Error>
+where
+    DB: BlockStore,
+{
+    let check_for_beacon_entry = |ts: &Tipset| {
+        let cbe = ts.blocks()[0].beacon_entries();
+        if let Some(entry) = cbe.last() {
+            return Ok(Some(entry.clone()));
+        }
+        if ts.epoch() == 0 {
+            return Err(Error::Other(
+                "made it back to genesis block without finding beacon entry".to_owned(),
+            ));
+        }
+        Ok(None)
+    };
+
+    if let Some(entry) = check_for_beacon_entry(ts)? {
+        return Ok(entry);
+    }
+    let mut cur = tipset_from_keys(db, ts.parents())?;
+    for _ in 1..20 {
+        if let Some(entry) = check_for_beacon_entry(&cur)? {
+            return Ok(entry);
+        }
+        cur = tipset_from_keys(db, cur.parents())?;
+    }
```

Comment:
Pretty minor, you could move the `cur = tipset_from_keys(db, cur.parents())?;` line to the top of the loop and initialise `cur` to just `ts`, this would save one `tipset_from_keys` call

Reply:
nah, because cur is owned, and ts is not. It looks a bit weird but cloning a tipset isn't trivial